File: kme/src/util/sampler.cpp
#include "sampler.h"

#include <chrono>
#include <boost/math/distributions/normal.hpp>

Sampler::Sampler(double seed) {
  if (seed == 0) {
    seed = std::chrono::system_clock::now().time_since_epoch().count();
  }
  rng = boost::mt19937(seed);
}

void Sampler::SampleNormalMatrix(MatrixXd *matrix, MatrixXd *means, double variance) {
  for (int i = 0; i < matrix->rows(); i++) {
    for (int j = 0; j < matrix->cols(); j++) {
      double mean = (means != NULL) ? (*means)(i, j) : 0.0;
      (*matrix)(i, j) = SampleNormal(mean, variance);
    }
  }
}

double Sampler::PdfNormal(double mean, double std, double x) {
  boost::math::normal_distribution<double> dist(mean, std);
  return boost::math::pdf<double>(dist, x);
}

double Sampler::SampleNormal(double mean, double std) {
  boost::normal_distribution<double> dist(mean, std);
  boost::variate_generator<boost::mt19937&, boost::normal_distribution<double>> sample(rng, dist);
  return sample();
}

double Sampler::SampleUniform(double min, double max) {
  boost::uniform_real<double> dist(min, max);
  boost::variate_generator<boost::mt19937&, boost::uniform_real<double>> sample(rng, dist);
  return sample();
}

int Sampler::SampleUniformInteger(int min, int max) {
  boost::uniform_int<int> dist(min, max);
  boost::variate_generator<boost::mt19937&, boost::uniform_int<int>> sample(rng, dist);
  return sample();
}

File: kme.py/util/sampler.py
import numpy as np
from scipy.stats import norm


class Sampler:
    def __init__(self, seed=None):
        if seed is None:
            seed = np.random.default_rng()
        self.rng = np.random.default_rng(seed)

    def sample_normal_matrix(self, shape, mean=0.0, variance=1.0):
        return self.rng.normal(loc=mean, scale=np.sqrt(variance), size=shape)

    def pdf_normal(self, mean, std, x):
        return norm.pdf(x, loc=mean, scale=std)

    def sample_normal(self, mean, std):
        return self.rng.normal(loc=mean, scale=std)

    def sample_uniform(self, low, high):
        return self.rng.uniform(low, high)

    def sample_uniform_integer(self, low, high):
        return self.rng.integers(low, high + 1)
